features/sketch/src/main/java/com/wire/android/feature/sketch/DrawingCanvasViewModel.kt

```diff
@@ -84,4 +105,63 @@ internal class DrawingCanvasViewModel : ViewModel() {
         )
     }
 
+    /**
+     * Sets the canvas size or modifies it if zoom is implemented.
+     */
+    fun onSizeChanged(canvasSize: Size) {
+        state = state.copy(canvasSize = canvasSize)
+    }
+
+    /**
+     * Undoes the last stroke.
+     */
+    fun onUndoLastStroke() {
+        if (state.paths.isNotEmpty()) {
+            state = state.copy(
+                paths = state.paths.dropLast(1),
+                pathsUndone = state.pathsUndone + state.paths.last()
+            )
+        }
+    }
+
+    /**
+     * Saves the image to the provided URI and resets the canvas.
+     *
+     * @param context The context to use to open the file descriptor.
+     * @param tempWritableImageUri The URI to save the image to.
+     *
+     * @return The [Uri] of the saved image.
+     */
+    suspend fun saveImage(context: Context, tempWritableImageUri: Uri?): Uri {
+        viewModelScope.launch {
+            withContext(Dispatchers.IO) {
+                with(state) {
+                    if (canvasSize == null || state.paths.isEmpty()) return@withContext
+
+                    val bitmap = Bitmap.createBitmap(
+                        canvasSize!!.width.toInt(),
+                        canvasSize!!.height.toInt(),
+                        Bitmap.Config.ARGB_8888
+                    )
+                    val canvas = Canvas(bitmap).apply { drawPaint(Paint().apply { color = Color.WHITE }) }
+                    context.contentResolver.openFileDescriptor(tempWritableImageUri!!, "rw")?.use { fileDescriptor ->
+                        FileOutputStream(fileDescriptor.fileDescriptor).use { fileOutputStream ->
+                            paths.forEach { path -> path.drawNative(canvas) }
+                            bitmap.compress(Bitmap.CompressFormat.JPEG, QUALITY, fileOutputStream)
+                            fileOutputStream.flush()
+                        }.also {
+                            Log.d("DrawingCanvasViewModel", "Image written to: $tempWritableImageUri")
+                        }
+                    }
+                }
+            }
+        }.join()
+        initializeCanvas()
+        return tempWritableImageUri!!
```

Comment:
Can't we also return a `Uri?` and handle if its nullable where else is being used? (in `onSendSketch` from `DrawingCanvasBottomSheet`)

Reply:
Good point, will give it a try sounds better to me :D

Reply:
Went for another option. The Uri should never be null, anyway, but, in case that happens, I'm just creating a new internal in cache now, so falling back even the first `kalium` managed Uri does not exist.